### DISCUSSION TITLE
block an IP if it has only some entry in the log but not others

### DISCUSSION THREAD

**Discussion by aalonzi** [fail2ban/fail2ban#3885] [Q&A] [answered]
I have a log file where I write the operations performed in a web application.

I would like to block an IP if it only log-in without doing anything else. Basically in the logs I have something like: 
```
[18-Nov-2024 10:03:01] [10.10.10.10] Username - IP in range 
[18-Nov-2024 10:03:02] [10.10.10.10] Username - Login Success 
[18-Nov-2024 10:04:01] [10.20.20.20] Username - IP in range 
[18-Nov-2024 10:04:02] [10.20.20.20] Username - Login Success 
[18-Nov-2024 10:04:03] [10.10.10.10] Username - IP in range 
[18-Nov-2024 10:04:04] [10.10.10.10] Username - Login Success
[18-Nov-2024 10:05:12] [10.20.20.20] Username - Save profile

```
In this case I would like to block the IP **10.10.10.10** but not the IP **10.20.20.20** (because it has a different entry than "**IP in range**" or "**Login Success**").

Is it possible to do it with fail2ban?

**Comment by sebres** [comment, marked as answer]
Strange approach... Something tells me also it may be incorrect from scratch.

Theoretically, it is indeed possible... Just unfortunately this log is insufficient for the decision point (when exactly it'd be considered as wanted to ban). It misses something like "Disconnected" (and it must be different message to others, because they will be marked as not a failure helper `<F-NOFAIL>` or success helper `<F-MLFGAINED>` and it'd need a different message to cause a fail state).
Therefore if one'd make the decision as the point of next login attempt (without "good" messages) instead of "Disconnected", it would not work (since it is no fail)...

In this case something like #1243 could help, but it is not fully implemented now for several reasons.

Fail2ban has also multi-line "mode" (`maxlines = N`), so one could write something like this:
```ini
maxlines = 10
failregex = ^(?:\[\])?\s*\[(?P<_caddr><ADDR>)\] <F-USER>\S+</F-USER> - Login Success[^\n]*<SKIPLINES>(?:^(?:\[\])?\s*\[(?P=_caddr)\] \S+ - (?!<F-NOFAIL>(?:IP in range|Login Success)[^\n]*</F-NOFAIL>)[^\n]*<SKIPLINES>)?(?=^(?:\[\])?\s*\[(?P=_caddr)\] \S+ - (?:IP in range|Login Success))
```
(to test it use `fail2ban-regex -L 10 -o "<ip> | <msg>" /path/to/log "$re_or_filter"`)

But it is ugly, slow and error-prone (would not work with some more in-between lines of other actors) and can be relative simply circumvented with small botnet. Let alone it'd expect 1 more "Login Success" message to consider a failures. So with above test excerpt it'd find only one failure of 10.10.10.10 (and the second will only counted if 3rd would follow, in maximal 10 lines).

**Reply by aalonzi** [reply]
Hi sebres, first of all thank you for the answer.

I'll try to add some details. Actually the logout message might not be there, because maybe the user didn't make a "logout" call and simply closed the browser.

_What I would like to implement is something like_:
If in 3 hours you have 6, or more, entries for a certain **username** like **Login Success** or **IP in range** but you don't have any other entries for this user (that is, there is no other line for that username that is different from **Login Success** or **IP in range** ), then you have to block it.

The goal is to avoid the use of automated systems that only log in without doing anything else (which only happens to check that the system is online, but this generates a series of complications).

**Reply by sebres** [reply]
>  Actually the logout message might not be there, because maybe the user didn't make a "logout" call and simply closed the browser.

It is not about logout, it may be end of request, disconnect or whatever message.
Otherwise it remains difficult to recognize the failure, or rather to distinguish the requests doing nothing after logon, or they being **not yet** doing something (but may do that later).
With other words "there is no other line for that username that is different from Login Success or IP in range" **yet** or **never**, so the time of end of work is not really defined.

Since you said "browser" - note that the work of browser may be more complex than you may assume - they could do some tasks via multiple connections (some of them can only log 2 types of those messages and nothing else), do a pre-caching for some URIs, etc.
So it'd introduce another additional factor for false positives.

As already said, it is anyway a questionable approach for my taste.

> The goal is to avoid the use of automated systems that only log in without doing anything else

1. Why such systems allowed to do `Login Success` and for them `IP in range` is then logged too?
Something is not well thought out by design.
2. And if all of them (everyone) really allowed, why it bothers at all (is-alive check is basically normal practice)?

> this generates a series of complications

Well, looks like a dirty workaround (instead of fixing this complications)...
For instance, if it is an issue of growing sessions (or those exhaustion either), then it'd be good to create the session by the first "task" (if user doing something) and not implicitly by connect/login.

Anyway - the filter I provided above may work, however with a restrictions I already mentioned (because of complexity with recognizing the end of work or else the timing issues).
